Add ede-dot-com tests that don't rely on internet name servers

divergentdave opened the pull request:
Closes #2491. This adds a separate Dockerfile that builds an older version of BIND, and sets up a Docker network that mirrors `extended-dns-errors.com` using modified versions of the same shell scripts. Separate tests are added to test recursors against subdomains hosted by these locally-run name servers. The offline-only tests are now run in CI, to catch future regressions.

bluejekyll commented:
Wow, this is a really great set of tests and configuration. Thanks for working on this @divergentdave

djc reviewed (approved):
This is great!

One doubt I have is about adding all the shell script. I personally don't have a lot of experience writing shell and so reviewing it is also not something I like spending time on. If there is an active upstream maintaining them (I see one "Major cleanup" from a months ago as the only maintainer commit since Sep 2023) there is some value in that, but here it feels like we're effectively forking them and maybe rewriting might be a feasible option with better long-term maintainability?

divergentdave commented on tests/ede-dot-com/src/lib.rs:
```diff
@@ -313,3 +578,143 @@ fn compare(subdomain: &str) -> Result<DigOutput> {
+        if line.trim().is_empty() {
+            continue;
+        }
+        let mut tokens = line.split_ascii_whitespace().collect::<Vec<_>>();
+        // Insert a TTL, since dnssec-signzone does not include one in DS records, and our FromStr
+        // implementations do not support this.
+        tokens.insert(1, "86400");
+        let ds: DS = tokens.join(" ").parse()?;
+        ds_records.push(ds);
```
As far as efficiency goes, `join()` calculates the exact size of the output before allocating once. Avoiding resizing the `String` might offset the extra allocation for collecting the vector in the first line. But I think we should keep this as-is mainly for clarity/simplicity.

divergentdave commented:
Rewriting would possible, but a fair bit more work. We'd probably need to improve multiple dns-test record types. I already had to track down several bugs where zones were not properly misconfigured due to my choice to use only one version of BIND, and I think the process reimplementing would introduce similar bugs.

djc commented:
Makes sense. I was hoping to mitigate that by rewriting the zone files from Rust while keeping the old code in place for a bit so that we can test for generating the exact same config/zones.

divergentdave commented:
Will do. I'm open to rewriting it, but I'd like to work on something else for a bit. Maybe once we've got all the RCODEs matching.

divergentdave commented:
A broad cross-section of CI jobs failed apparently due to connection issues with Google Public DNS. I'm going to wait a bit and try again.

divergentdave commented:
Belated follow-up: I just noticed the news section of https://extended-dns-errors.com/ lists the changes made to the available subdomains. This explains the differences from the list in the IMC 2023 paper.

> [08.10.2024] Major cleanup: 1) All the domains (apart from nsec3-iter-*) are now using 0 additional iterations and no salt; 2) Added more domains that are not compliant with RFC 9276; 3) Removed most of the domains with non-routable glue records due to redundant results.
